modules/core/src/main/java/org/apache/ignite/internal/util/GridConcurrentMultiPairQueue.java

```diff
@@ -0,0 +1,162 @@
+ * and further sequence of {@code poll} calls may produce output like: <br>
+ * [3, 200], [3, 100], [1, 1], [1, 3], [1, 5], [1, 7], [2, 2], [2, 3]
+ *
+ * @param <K> The type of key in input pair collection.
+ * @param <V> The type of value array.
+ */
+public class GridConcurrentMultiPairQueue<K, V> {
+    /** */
+    public static final GridConcurrentMultiPairQueue EMPTY =
+        new GridConcurrentMultiPairQueue<>(Collections.emptyMap());
+
+    /** Inner holder. */
+    private final V[][] vals;
+
+    /** Storage for every array length. */
+    private final long[] lenSeq;
+
+    /** Current absolute position. */
+    private final AtomicLong pos = new AtomicLong();
+
+    /** Precalculated max position. */
+    private final int maxPos;
+
+    /** Keys array. */
+    private final K[] keysArr;
+
+    /** Thread local cached segment. */
+    private ThreadLocal<Integer> cachedSegment = ThreadLocal.withInitial(() -> 0);
+
+    /** */
+    public GridConcurrentMultiPairQueue(Map<K, ? extends Collection<V>> items) {
+        int pairCnt = (int)items.entrySet().stream().map(Map.Entry::getValue).filter(k -> k.size() > 0).count();
+
+        vals = (V[][])new Object[pairCnt][];
+
+        keysArr = (K[])new Object[pairCnt];
+
+        lenSeq = new long[pairCnt];
+
+        int keyPos = 0;
+
+        int size = -1;
+
+        for (Map.Entry<K, ? extends Collection<V>> p : items.entrySet()) {
+            if (p.getValue().size() == 0)
+                continue;
+
+            keysArr[keyPos] = p.getKey();
+
+            lenSeq[keyPos] = size += p.getValue().size();
+
+            vals[keyPos++] = (V[])p.getValue().toArray();
+        }
+
+        maxPos = size + 1;
+    }
+
+    /** */
+    public GridConcurrentMultiPairQueue(Collection<T2<K, V[]>> items) {
+        int pairCnt = (int)items.stream().map(Map.Entry::getValue).filter(k -> k.length > 0).count();
+
+        vals = (V[][])new Object[pairCnt][];
+
+        keysArr = (K[])new Object[pairCnt];
+
+        lenSeq = new long[pairCnt];
+
+        int keyPos = 0;
+
+        int size = -1;
+
+        for (Map.Entry<K, V[]> p : items) {
+            if (p.getValue().length == 0)
+                continue;
+
+            keysArr[keyPos] = p.getKey();
+
+            lenSeq[keyPos] = size += p.getValue().length;
+
+            vals[keyPos++] = p.getValue();
+        }
+
+        maxPos = size + 1;
+    }
+
+    /**
+     * Retrieves and removes the head of this queue,
+     * or returns {@code null} if this queue is empty.
+     *
+     * @return the head of this queue, or {@code null} if this queue is empty
+     */
+    public @Nullable T2<K, V> poll() {
+        long absPos = pos.getAndIncrement();
+
+        if (absPos >= maxPos)
+            return null;
+
+        int segment = cachedSegment.get();
+
+        if (absPos > lenSeq[segment]) {
+            segment = Arrays.binarySearch(lenSeq, absPos);
+
+            segment = segment < 0 ? -segment - 1 : segment;
```

Comment:
I have not read your tests yet. Does this particular line has its own test case?

Reply:
-segment - 1 from bsearch documentation shows that input key not exactly belongs to array but nested between to indexes.

Reply:
I know that) What I meant is that it's easy to miss some "+1" or "-1" in these tricky expressions, I'm always a bit paranoid about such errors.
I rechecked this line with paper and pen, it's correct.